### ISSUE TITLE
Sending string between address spaces on AMD64

### ISSUE THREAD

**Issue by mandrookin** [l4ka/pistachio#5]
Kernel goes to an infinity loop when sending string between address spaces on AMD64

Problem detected around line 406 of pistachio/kernel/src/glue/v4-x86/space.cc 

```
enter_kdebug("ENTER: dst_pgent  ->set_entry");
dst_pgent  ->set_entry (this, pgsize, *src_pgent  );
enter_kdebug("LEAVE: dst_pgent  ->set_entry");
```

Debugging steps between "ENTER" and "LEAVE" avoids the problem until next call. 
Continue running on "ENTER" will cause kernel to go to an infinity loop.

Problem has been reproduced under VMWare.

config.h:

/\* Linux version 2.6.37.6 (root@midas64) (gcc version 4.5.2 (GCC) ) #3 SMP Sat Apr 9 22:49:32 CDT 2011 */
# define CONFIG_ARCH_X86 1
# define CONFIG_SUBARCH_X64 1
# define CONFIG_CPU_X86_K8 1
# define CONFIG_PLAT_PC99 1
# define CONFIG_EXPERIMENTAL 1
# define CONFIG_X_PAGER_EXREGS 1
# define CONFIG_SCHED_RR 1
# define CONFIG_DEBUG 1
# define CONFIG_K8_FLUSHFILTER 1
# define CONFIG_NEW_MDB 1
# define CONFIG_STATIC_TCBS 1
# define CONFIG_KDB_DISAS 1
# define CONFIG_KDB_ON_STARTUP 1
# define CONFIG_KDB_INPUT_HLT 1
# define CONFIG_X86_PSE 1
# define CONFIG_X86_SYSENTER 1
# define CONFIG_X86_PGE 1
# define CONFIG_X86_FXSR 1
# define CONFIG_X86_PAT 1
# define CONFIG_IS_64BIT 1
# define CONFIG_X86_TSC 1
# define AUTOCONF_INCLUDED

**Comment by jstoess** [comment]
Can you do me a favor and retry with the following options DISABLED:

#define CONFIG_K8_FLUSHFILTER 1

#define CONFIG_STATIC_TCBS 1

#define CONFIG_NEW_MDB 1

thanks.

**Comment by mandrookin** [comment]
Sorry for response delay. Here is some report on problem

#undef CONFIG_NEW_MDB cause build errors:

./l4ka-pistachio-5c1b29b/kernel/src/generic/mdb_mem.cc:198:64: error: no matching function for call to 'pgent_t::vaddr(space_t_, pgent_t::pgsize_e, mdb_node_t_&)'
./l4ka-pistachio-5c1b29b/kernel/src/generic/mdb_mem.cc:239:64: error: no matching function for call to 'pgent_t::vaddr(space_t_, pgent_t::pgsize_e, mdb_node_t_&)'
./l4ka-pistachio-5c1b29b/kernel/src/generic/mdb_mem.cc:318:64: error: no matching function for call to 'pgent_t::vaddr(space_t_, pgent_t::pgsize_e, mdb_node_t_&)'
./l4ka-pistachio-5c1b29b/kernel/src/generic/mdb_mem.cc:330:64: error: no matching function for call to 'pgent_t::vaddr(space_t_, pgent_t::pgsize_e, mdb_node_t_&)'
./l4ka-pistachio-5c1b29b/kernel/src/generic/mdb_mem.cc:358:45: error: no matching function for call to 'pgent_t::vaddr(space_t_&, pgent_t::pgsize_e&, mdb_node_t_&)'

#undef CONFIG_STATIC_TCBS - kernel locks with 100% CPU load on starting.

#undef CONFIG_K8_FLUSHFILTER - do not change behavior. 
I tested IPC with help of user/apps/l4test. Only "Simple IPC" test passed. IPC with strings go to an infinity loop.

Should I try some specific gcc version? I am using Slackware 64-bit build 13.37.0

**Comment by jstoess** [comment]
> #undef CONFIG_NEW_MDB cause build errors:
> #undef CONFIG_STATIC_TCBS - kernel locks with 100% CPU load on starting.

 I cannot verify this. I can compile without MDB, using the most recent version on github. Also I'm almost always undefing static tcbs.
You better turn static TCBs as well. Using static TCBs only makes sense on embedded architectures with limited virtual address space. Non-static TCBs were default for years, by the way, and the static TCB implementation is way newer.

I suspect something else might be broken on your side, have a look at 
http://www.l4ka.org/98.php
where we test L4 configurations automatically, using qemu. You see some of them failing but not those related to your config (basically all configs except x86-x64-*.statictcbs are 64-bit versinos compiled without CONFIG_STATIC_TCBS, and most of them do well.

Are you using stock L4Ka::Pistachio? I don't have Slackware, which toolchain are you using? GCC compiler

**Comment by mandrookin** [comment]
Thank you very much for assistance. I tested recent version L4Ka::Pistachio which has been compiled on Suse 11.10. It works perfect in VirtualBox. 

Also I tested same bootable image in VMWare and found that VMWare is not compatible with x86-x64 version of Pistachio:
- it goes to an infinity loop after "Adding more kernel memory" message. 
- when I comment add_more_kmem(), then kernel goes to the infinity loop in init_kernel_threads() of api/v4/thread.cc

So it is not 64-bit issue but VMWare issue. Should I close this issue or it could be fixed some day?
